include/cuco/probing_scheme.cuh

```diff
@@ -175,6 +177,159 @@ class double_hashing : private detail::probing_scheme_base<CGSize> {
   Hash2 hash2_;
 };
 
+/**
+ * @brief wrapper to test if a probing scheme is perfect
+ * @tparam T is a probing scheme that does not have a method "is_perfect".
+ **/
+template <typename T, typename = void>
+struct is_perfect_hashing : cuda::std::false_type {};
+
+/**
+ * @brief wrapper to test if a probing scheme is perfect
+ * @tparam T is a probing scheme that does have a method "is_perfect".
+ **/
+template <typename T>
+struct is_perfect_hashing<T, cuda::std::void_t<decltype(std::declval<T>().is_perfect())>>
+  : cuda::std::true_type {};
+
+/**
+ * @brief Public perfect probing scheme class.
+ *
+ * @note Perfect hash functions guarantee no collisions. User is responsible for supplying a perfect
+ * hash
+ *
+ * @note User must ensure that the Hash function in combination with the input key set actually
+ * forma a perfect hash function
+ *
+ * @note User must ensure that the maximum hash values is smaller than the map's capacity
+ *
+ * @note `Hash` should be callable object type.
+ *
+ * @tparam CGSize Size of CUDA Cooperative Groups
+ * @tparam Hash Unary callable type
+ */
+template <int32_t CGSize, typename Hash>
+class perfect_probing : private detail::probing_scheme_base<CGSize> {
+ public:
+  using probing_scheme_base_type =
+    detail::probing_scheme_base<CGSize>;  ///< The base probe scheme type
+  using probing_scheme_base_type::cg_size;
+  static bool constexpr is_perfect() { return true; }
+
+  /**
+   *@brief Constructs perfect probing scheme with the hasher callable.
+   *
+   * @param hash Hasher
+   */
+  __host__ __device__ constexpr perfect_probing(Hash const& hash = {});
+
+  /**
+   *@brief Makes a copy of the current probing method with the given hasher
+   *
+   * @tparam NewHash New hasher type
+   *
+   * @param hash Hasher
+   *
+   * @return Copy of the current probing method
+   */
+  template <typename NewHash>
+  [[nodiscard]] __host__ __device__ constexpr auto with_hash_function(
+    NewHash const& hash) const noexcept;
```

Comment:
Also here, static_assert that the new hash function is perfect.  this will *not* catch users that provide a perfect hash function that has a range which is bigger than the number of slots in the hash table!  That's up to the user to be smart enough to do that properly.  We should probably put it in the doxygen that the hash function should have a range smaller than the number of slots in the hash table and it's up to the user to ensure that.  We can't ensure on the user's behalf.

We could do some tricks to help the user out.  For example, we could have the user provide, along with the hash, the maximum possible value of the hash.  And then we would do a one-time check on the CPU to see that the maximum possible output of the hash is smaller than the number of slots.  This is just taking a very small check that the user is going to have to do anyway and forcing it to happen by forcing the user to provide us with the inputs to do it.  Is that a good idea?  I dunno, maybe?  It may prevent user errors.  And the user is going to have to know that maximum value anyway in order to construct a hash table that is large enough.

If you want, this could be a later refinement.  It would cause the user to get a runtime error explaining that perfect probing was used incorrectly, rather than the code silently corrupting memory.

In my implementation, in addition to the `is_perfect` method, I have a `max()` method that returns the maximum possible output of a hash function.  I use that metaprogramming trick from above to detect if a hasher has a `max()` method.  If the method exists, return that value.  If it does not then return [`std::numeric_limits<T>::max`](https://en.cppreference.com/w/cpp/types/numeric_limits/max).  `max()` is not constexpr.  Maybe call it `range_max` to differentiate it from being interpreted as the max of the domain?

Reply:
I do like this idea, and the details added in your other [comment](Wait, no, the answer is that perfect probing doesn't need CGSize because it will always be 1. Remove it as a template param and get rid of all mention of cooperative groups in perfect probing.).